### PR TITLE
Implement BINARY_SUBSCR specialization for (String Int)

### PR DESCRIPTION
**Pull request by corona10**
[Failure ratio](https://github.com/faster-cpython/ideas/blob/841a673b66e3089257813fc6bb33e5c87afd0ffa/stats.md): 24.2%(string int) from `BINARY_SUBSCR`

## Microbenchmark

Mean +- std dev: [base] 3.75 us +- 0.02 us -> [BINARY_SUBSCR_UNICODE_INT] 3.48 us +- 0.02 us: 1.08x faster
`````python
import pyperf

runner = pyperf.Runner()
runner.timeit(name="str index",
              stmt="""
i = 0
while i < len(s):
    ch = s[i]
    i += 1
""",
              setup = """
s = 'helloworld' * 10
"""
              )

`````

## Ref leak test
````

➜  cpython git:(BINARY_SUBSCR_UNICODE_INT) ✗ ./python.exe -m test test_unicode test_userstring test_bytes -R 3:3
Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 3.51 Run tests sequentially
0:00:00 load avg: 3.51 [1/3] test_unicode
beginning 6 repetitions
123456
......
0:00:09 load avg: 3.12 [2/3] test_userstring
beginning 6 repetitions
123456
......
0:00:17 load avg: 2.95 [3/3] test_bytes
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

All 3 tests OK.

Total duration: 32.2 sec
Tests result: SUCCESS
````

### CODE REVIEW

**Comment by corona10** [comment]
@kumaraditya303 you are right, specializing in ASCII only shows better performance.
I will take a look at the coverage of BINARY_SUBSCR_ASCII_INT for BINARY_SUBSCR_STR_INT
Thank you for the suggestion and please take a look at that whether I handle carefully ascii string look up :)

## Microbenchmark
`````

Mean +- std dev: [base] 3.75 us +- 0.02 us -> [BINARY_SUBSCR_ASCII_INT] 3.36 us +- 0.03 us: 1.12x faster
`````

**Comment by corona10** [comment]
Here is stat: https://gist.github.com/corona10/bdf7768e7d634dca87034f1f9506abf3

## AS-IS

### BINARY_SUBSCR

|Kind | Count | Ratio | 
|---|---|---|
|  unquickened |     26777990 | 1.0% |
| specialization.deferred |   1099812093 | 41.2% |
| specialization.deopt |       400664 | 0.0% |
|          hit |   1520768373 | 57.0% |
|         miss |     21378050 | 0.8% |

#### Specialization attempts

| | Count | Ratio | 
|---|---:|---:|
| Success | 745,889 | 45.8% |
| Failure | 882,644 | 54.2% |

|Failure kind | Count | Ratio | 
|---|---:|---:|
| string int | 213,383 | 24.2% |
| string slice | 175,312 | 19.9% |
| array int | 166,815 | 18.9% |
| list slice | 101,907 | 11.5% |
| buffer slice | 92,514 | 10.5% |
| other | 79,465 | 9.0% |
| buffer int | 32,637 | 3.7% |
| tuple slice | 17,103 | 1.9% |
| sequence int | 3,262 | 0.4% |
| array slice | 246 | 0.0% |

## TO-BE

### BINARY_SUBSCR
|Kind | Count | Ratio |
|---|---|---|
|  unquickened |     43401421 | 1.1% |
| specialization.deferred |   1442458722 | 36.8% |
| specialization.deopt |       685055 | 0.0% |
|          hit |   2400478471 | 61.2% |
|         miss |     36691726 | 0.9% |

#### Specialization attempts

| | Count | Ratio |
|---|---:|---:|
| Success | 1,135,256 | 55.5% |
| Failure | 909,350 | 44.5% |

|Failure kind | Count | Ratio |
|---|---:|---:|
| string slice | 243,475 | 26.8% |
| array int | 225,573 | 24.8% |
| list slice | 132,861 | 14.6% |
| buffer slice | 118,608 | 13.0% |
| other | 107,661 | 11.8% |
| buffer int | 47,972 | 5.3% |
| tuple slice | 26,443 | 2.9% |
| array slice | 3,469 | 0.4% |
| sequence int | 3,288 | 0.4% |

File: Python/ceval.c [hunk 1]
```diff
@@ -2207,6 +2208,34 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+
+            // Deopt unless 0 <= sub < ascii->length
+            PyASCIIObject *ascii = _PyASCIIObject_CAST(unicode);
+            Py_ssize_t signed_magnitude = Py_SIZE(sub);
+            DEOPT_IF(((size_t)signed_magnitude) > 1, BINARY_SUBSCR);
+            assert(((PyLongObject *)_PyLong_GetZero())->ob_digit[0] == 0);
+            Py_ssize_t index = ((PyLongObject*)sub)->ob_digit[0];
+            DEOPT_IF(index >= ascii->length, BINARY_SUBSCR);
+            STAT_INC(BINARY_SUBSCR, hit);
+            Py_UCS4 ch = PyUnicode_1BYTE_DATA(ascii)[index];
```
Comment (kumaraditya303) [inline Python/ceval.c]: This still has branches as it checks if the string is ascii or not. The following should be much faster.
```suggestion
            const void *data = (void*)(ascii + 1);
            Py_UCS1 ch = ((Py_UCS1*)data)[index];
```

**Comment by corona10** [comment]
With @kumaraditya303 's suggestion:
Mean +- std dev: [base] 3.82 us +- 0.06 us -> [opt] 3.28 us +- 0.02 us: 1.17x faster

File: Python/ceval.c [hunk 2]
```diff
@@ -2248,6 +2249,34 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+
+            // Deopt unless 0 <= sub < ascii->length
+            PyASCIIObject *ascii = _PyASCIIObject_CAST(unicode);
+            Py_ssize_t signed_magnitude = Py_SIZE(sub);
+            DEOPT_IF(((size_t)signed_magnitude) > 1, BINARY_SUBSCR);
+            assert(((PyLongObject *)_PyLong_GetZero())->ob_digit[0] == 0);
+            Py_ssize_t index = ((PyLongObject*)sub)->ob_digit[0];
+            DEOPT_IF(index >= ascii->length, BINARY_SUBSCR);
+            STAT_INC(BINARY_SUBSCR, hit);
+            Py_UCS1 ch = *((Py_UCS1*)(ascii + 1) + index);
```
Comment (markshannon) [inline Python/ceval.c:2268]: This is a bit obscure. Is there a pre-existing macro for getting the first character in an ascii string?

File: Python/ceval.c [hunk 3]
```diff
@@ -2248,6 +2249,34 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+            PyObject *res = (PyObject*)&_Py_SINGLETON(strings).ascii[ch];
```
Comment (markshannon) [inline Python/ceval.c]: Also a bit obscure. It might be worth adding macros to `pycore_unicodeobject.h`

**Review by markshannon — COMMENTED** [review]
A couple of niggles.
Do you have pyperformance results?

**Comment by corona10** [comment]
> A couple of niggles.
> Do you have pyperformance results?

Sorry, I don't have a reliable isolated environment to run pyperformance.
But I can run benchmarks with my m1 mac machine (but not isolated..)
or If you don't mind, would you like to run a benchmark for this PR? 
Sorry for the frequent request.

File: Python/ceval.c (hunk 2 above)
Comment (corona10) [reply Python/ceval.c:2268]: Not exist but I am not sure about adding the macro for this.

**Comment by markshannon** [comment]
[A bit faster, but not significantly](https://gist.github.com/markshannon/0ac4116b80b96945e354d52dfd1e1e98)

**Comment by corona10** [comment]
> A bit faster, but not significantly

Thanks!
Should we satisfy both performance improvement and high failure ratio reduction in the entire benchmark set?

**Comment by erlend-aasland** [comment]
I fail to see how this qualifies as _skip issue_, according to our workflow.

**Comment by corona10** [comment]
> I fail to see how this qualifies as skip issue, according to our workflow

I will create a issue for the future work, but let's close this PR now.

**Pull request closed without merging**